### PR TITLE
Bare bones for differential fuzzing

### PR DESCRIPTION
**Pull request by satyamakgec**
### What was wrong?

Needs fuzzing of contracts.

### How was it fixed?

It consists of the fuzzing setup for the erc20.fe and erc20.sol and has the rearrangements for the compiler tests as an independent directory to avoid circular dependency and it will also decrease the test execution time as it doesn't require rebuilding the dependencies again.

To run the fuzzer - `cd fuzz && cargo +nightly fuzz run erc20_target --features solc-backend`

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [ ] OPTIONAL: Update [Spec](https://github.com/ethereum/fe/blob/master/docs/src/spec/index.md) if applicable
- [ ] Add entry to the [release notes](https://github.com/ethereum/fe/blob/master/newsfragments/README.md) (may forgo for trivial changes)

- [ ] Clean up commit history

### CODE REVIEW

**Comment by sbillig** [comment]
> ohh sorry @sbillig I should have mention this in the readme. It only works with the feature flag i.e solc-backend

Ha! Somehow I didn't think of that, despite using `--all-features` all day long to run other tests. Thanks.

**Review by cburgdorf — COMMENTED** [review]
Excited to see it coming together. Sorry, not a thorough review, I can't squeeze anymore out of the day. Counting on more pairs of :eyes: to go over it within the next few hours anyway.

File: fuzz/src/erc20.rs [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+}
+
+pub fn fe_erc20_transfer(info: TransferInfo) {
+    if info.value >= 1000000000000000000000000 as u128 {
+        with_executor(&|mut executor| {
+            let mut harness = deploy_contract(
+                &mut executor,
+                "test-utils/fuzz/fixtures/fe/erc20.fe", // TODO: Need to fix the path
+                "ERC20",
+                &[string_token("Fe Coin"), string_token("FE")],
```
Comment (g-r-a-n-t) [inline fuzz/src/erc20.rs:58]: This is a good start, but it doesn't test much.

I think what we eventually want is to have two separate deployed contracts/executors that we perform many state-modifying transactions on in a random order. This will get us into edge cases where we might find bugs.

Of course this doesn't need to be included in this PR, I'm just wondering if you've been thinking about this.

**Review by g-r-a-n-t — APPROVED** [review]

File: fuzz/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,30 @@
+[package]
+authors = ["Automatically generated"]
+edition = "2018"
+name = "fe-compiler-fuzz"
```
[the remark this answers is not in the document]
Comment (cburgdorf) [reply fuzz/Cargo.toml:4]: >I think we should disambiguate it from the fuzzing that Alex is doing, which I would call compiler fuzzing.

Ah, right, makes sense. I was looking for something that was a single word and lines up with the directory name but it's not a big deal either way.

[the remark this answers is not in the document]
Comment (g-r-a-n-t) [reply fuzz/Cargo.toml:4]: Yeah, `fe-fuzz` or `fe-fuzzer` makes sense. I guess this could contain more than differential fuzzing, so `fe-diff-fuzz` is probably too specific.

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -2,169 +2,167 @@ name: CI
...
+  #     steps:
+  #     - uses: actions/checkout@v2
+  #     - name: Install rust
+  #       uses: actions-rs/toolchain@v1
+  #       with:
+  #         profile: minimal
+  #         toolchain: stable
+  #         override: true
+  #     - name: Run WASM tests
+  #       run: wasm-pack test --node -- --workspace
```
[the remark this answers is not in the document]
Comment (satyamakgec) [reply .github/workflows/main.yml:128]: This is intentional because now tests are living in a separate library which needs to go under workspace so new changes lead to compile the solc bindings to wasm as well which is not possible right now as `wasm32-unknown-unknown` is not supporting c++ bindings right now. We did decided to comment the wasm-pack test cases on the CI and to make it work for now and for the releases we will manually run the wasm-pack test cases locally to verify whether everything is good or not.

File: fuzz/src/erc20.rs [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+                "ERC20",
+                &[string_token(details.name), string_token(details.symbol)],
+            );
+            erc20_constructor(details, &mut executor, harness);
+        });
+    }
+}
+
+pub fn fe_erc20_transfer(info: TransferInfo) {
+    if info.value >= 1000000000000000000000000 as u128 {
```
[the remark this answers is not in the document]
Comment (satyamakgec) [reply fuzz/src/erc20.rs]: Actually, it should be <= so use shouldn't be able to transfer more than the total supply.

File: test-utils/src/lib.rs [hunk 5]
```diff
@@ -301,7 +306,21 @@ fn _deploy_contract(
     panic!("Failed to create contract")
 }
 
-pub fn compile_solidity_contract(name: &str, solidity_src: &str) -> Result<(String, String), ()> {
+#[derive(Debug)]
+pub struct NotAbleToCompile;
```
[the remark this answers is not in the document]
Comment (satyamakgec) [reply test-utils/src/lib.rs]: Done, I added the vector of errors to know all the errors at once. Skipping warnings though

File: fuzz/Cargo.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (satyamakgec) [reply fuzz/Cargo.toml]: `fe-fuzzer` get choosen

File: fuzz/src/erc20.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (g-r-a-n-t) [reply fuzz/src/erc20.rs]: The expected behavior is that the tx reverts, correct? I think we should be validating this.

File: fuzz/src/erc20.rs [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+            erc20_transfer(info, executor, harness);
+        });
+    }
+}
+
+pub fn sol_erc20_transfer(info: TransferInfo) {
+    if info.value
+        <= TOTAL_SUPPLY
+            .parse::<u128>()
+            .unwrap_or(1000000000000000000000000)
```
Comment (g-r-a-n-t) [inline fuzz/src/erc20.rs:73]: We should just `expect` here or make `TOTAL_SUPPLY` a `u128`.

Comment (satyamakgec) [reply fuzz/src/erc20.rs:73]: Okay, will add that change in another PR.

File: fuzz/src/erc20.rs (hunk 1 above)
Comment (satyamakgec) [reply fuzz/src/erc20.rs:62]: Yeah, I was thinking about this to have some kind of random order testing. I have a process in mind where we can come up with some mutation testing where we can randomly mutate the contract as well if we want but that is I think one step forward but for now, I am thinking to have some kind of prop testing with random ordering.
